File: src/presentation/mcp/toolsets/system.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# SPDX-FileCopyrightText: 2026 Greenbone AG

"""MCP tools for system operations."""


from typing import TYPE_CHECKING, Any

from src.services.system import SystemService

if TYPE_CHECKING:
    from mcp.server.fastmcp import FastMCP


def register_system_tools(server: FastMCP, service: SystemService) -> None:
    """Register system tools with MCP server.

    Args:
        server: FastMCP server instance.
        service: System service instance.
    """

    @server.tool(name="openvas_get_version")
    def get_version() -> dict[str, Any]:
        """Get GVM version information.

        Returns:
            GVM version details including GMP protocol version.
        """
        result = service.get_version()
        return result.model_dump()

    @server.tool(name="openvas_test_connection")
    def test_connection() -> dict[str, Any]:
        """Test connection to GVM server.

        Returns:
            Connection status and version if connected.
        """
        try:
            version = service.get_version()
            return {
                "connected": True,
                "gmp_version": version.gmp_version,
            }
        except Exception as e:
            return {
                "connected": False,
                "error": str(e),
            }

File: src/mcp/tools/system.rs
//! System toolset: connectivity and version tools.

use rmcp::model::{CallToolResult, ContentBlock};
use rmcp::{ErrorData as McpError, tool, tool_router};
use serde::Serialize;

use crate::gateway::models::VersionInfo;
use crate::mcp::error::gateway_tool_error;
use crate::mcp::server::GvmMcpServer;

use super::common::json_result;

/// Structured payload returned by `openvas_test_connection`.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct TestConnectionReport {
    gateway_url: String,
    gateway_status: String,
    api_version: String,
    gmp_version: String,
    /// Whether the identity used for this call is accepted by the gateway.
    authenticated: bool,
    read_only: bool,
    toolsets: String,
}

#[tool_router(router = system_router, vis = "pub(crate)")]
impl GvmMcpServer {
    /// Verify connectivity to the GVM stack: checks gateway liveness, queries
    /// the gvmd version and makes one authenticated call to confirm the
    /// forwarded identity is accepted. Use this first if other tools fail.
    #[tool(
        name = "openvas_test_connection",
        annotations(title = "Test GVM connection", read_only_hint = true)
    )]
    pub async fn test_connection(&self) -> Result<CallToolResult, McpError> {
        let health = match self.gateway().health().await {
            Ok(health) => health,
            Err(err) => {
                return Ok(gateway_tool_error(
                    "checking gateway liveness (GET /health)",
                    &err,
                ));
            }
        };

        let version: VersionInfo = match self.gateway().version().await {
            Ok(version) => version,
            Err(err) => {
                return Ok(gateway_tool_error(
                    "querying gvmd version (GET /api/v1/version)",
                    &err,
                ));
            }
        };

        // One authenticated call verifies the identity the gateway will use
        // for real tool calls (the caller's forwarded credentials, or the
        // configured fallback). A 401 here surfaces as "authentication failed".
        if let Err(err) = self
            .gateway()
            .get_json_query::<serde_json::Value>(&["targets"], &[("perPage", "1".to_string())])
            .await
        {
            return Ok(gateway_tool_error(
                "verifying gateway credentials (GET /api/v1/targets)",
                &err,
            ));
        }

        let report = TestConnectionReport {
            gateway_url: self.config().gateway_url.to_string(),
            gateway_status: health.status,
            api_version: version.api_version,
            gmp_version: version.gmp_version,
            authenticated: true,
            read_only: self.config().read_only,
            toolsets: self.config().toolsets.to_string(),
        };

        let text = serde_json::to_string_pretty(&report)
            .map_err(|err| McpError::internal_error(err.to_string(), None))?;
        Ok(CallToolResult::success(vec![ContentBlock::text(text)]))
    }

    /// Get the GVM REST API contract version and the GMP protocol version
    /// reported by gvmd.
    #[tool(
        name = "openvas_get_version",
        annotations(title = "Get GVM versions", read_only_hint = true)
    )]
    pub async fn get_version(&self) -> Result<CallToolResult, McpError> {
        match self.gateway().version().await {
            Ok(version) => json_result(&version),
            Err(err) => Ok(gateway_tool_error(
                "querying gvmd version (GET /api/v1/version)",
                &err,
            )),
        }
    }
}
